python/graphstorm/model/lp_gnn.py

```diff
@@ -107,6 +109,100 @@ def forward(self, blocks, pos_graph,
         return pred_loss + alpha_l2norm * reg_loss
 
+
+def get_embs(emb, node_list, neg_sample_type, canonical_etype):
+    """ Fetch node embeddings for mini-batch prediction.
+
+    Parameters
+    ----------
+    emb: dict of Tensor
+        Node embeddings.
+    node_list: tuple of tensor
+        list of positive/negative source and destination nodes
+    neg_sample_type: str
+        Describe how negative samples are sampled.
+            Uniform: For each positive edge, we sample K negative edges
+            Joint: For one batch of positive edges, we sample
+                   K negative edges
+    canonical_etype: str
+        Relation type
+
+    Return
+    ------
+    tuple of tensors
+        node embeddings stored in a tuple:
+        tuple(positive source embeddings, negative source embeddings,
+        postive destination embeddings, negatve destination embeddings).
+    """
+    utype, _, vtype = canonical_etype
+    pos_src, neg_src, pos_dst, neg_dst = node_list
+    pos_src_emb = emb[utype][pos_src]
+    pos_dst_emb = emb[vtype][pos_dst]
+    neg_src_emb = None
+    neg_dst_emb = None
+
+    if neg_src is not None:
+        neg_src_emb = emb[utype][neg_src.reshape(-1,)]
+        if neg_sample_type == BUILTIN_LP_UNIFORM_NEG_SAMPLER:
+            neg_src_emb = neg_src_emb.reshape(neg_src.shape[0], neg_src.shape[1], -1)
+    if neg_dst is not None:
+        if neg_sample_type == BUILTIN_LP_UNIFORM_NEG_SAMPLER:
+            neg_dst_emb = emb[vtype][neg_dst.reshape(-1,)]
+            neg_dst_emb = neg_dst_emb.reshape(neg_dst.shape[0], neg_dst.shape[1], -1)
+        elif neg_sample_type == BUILTIN_LP_JOINT_NEG_SAMPLER:
+            neg_dst_emb = emb[vtype][neg_dst]
+        else:
+            assert False, f"Unknow negative sample type {neg_sample_type}"
+    return (pos_src_emb, neg_src_emb, pos_dst_emb, neg_dst_emb)
+
+
+def cat_nodelist_of_multiple_batches(pos_neg_tuple, loader, num_batch_to_cat):
+    """ Concatenate node lists from multiple batches and retrieve their embeddings
+        with a single remote pull.
+
+    Parameters
+    ----------
+    pos_neg_tuple: dict of tuple
+        Positive and negative edges stored in a tuple:
+        tuple(positive source, negative source,
+        postive destination, negatve destination).
+        The positive edges: (positive source, positive desitnation)
+        The negative edges: (positive source, negative desitnation) and
+                            (negative source, positive desitnation)
+    loader : GSgnnEdgeDataLoader
+        The GraphStorm dataloader
+    num_batch_to_cat: int
+        The number of batches aka pos_neg_tuple to concatenate which
+        will be used to fetch embeddings in get_embs() function
+
+    Return
+    ------
+    node_list: Dict of tuple of tensors
+        Dictionary of edge type to concatenated node list stored in
+        a tuple
+    """
+    node_list = {}
+    canonical_etype = prev_canonical_etype = list(pos_neg_tuple.keys())[0]
```

Comment:
Why we need canonical_etype and prev_canonical_etype? They are pointing to the same item.